content/blog/2024-08-01-gsoc-multiboot2.mdx

```diff
@@ -0,0 +1,98 @@
+
+> With the Multiboot2 header issue resolved, the ELF file is now generated correctly.
+
+Boy, was I wrong!
+The past three weeks have been a rollercoaster of debugging, testing, and more debugging.
+That being said, considerable progress has been made, which will be the focus of this update post.
+
+## The Debugging Chronicles
+
+In reviewing the issues with the infinite loop, I realized that the root cause was the omission of the end tag in the Multiboot2 header.
+Multiboot2 headers require a specific ending structure to signal the completion of the header sequence.
+Without this tag, the bootloader, GRUB in this case, continues to search for it indefinitely, leading to the infinite loop.
+Now that the issue was resolved, a new contender emerged: the "out of memory" error.
+Let the games begin!
+
+### The "Out of Memory" Error
+
+The "out of memory" error was a puzzling one.
+The first instinct was to increase the memory size, but no matter how much memory was allocated, the error persisted.
+This meant that either something was null or the area where the memory was being allocated was already populated due to possible fragmentation.
+
+Since the error was occuring in the `grub_relocator_alloc_chunk_align` function, I decided to find its caller to reconstitute the flow and logic leading here.
+There were several difficulties with debugging this.
+
+Firstly, I was connected to a 64-bit `gdb` server, but the error was occurring in 32-bit mode so the registers were all messed up.
+To make life a bit easier, I ended up using `qemu-system-i386` instead of `qemu-system-x86_64`.
+Sure, I couldn't boot Unikraft like this, but I could at least debug the issue at hand.
+
+Additionally, I had no symbols.
+Therefore, I had to load them manually using `add-symbol-file` in `gdb` for each module I was interested in.
+
+It is important to mention what I mean by "module".
+GRUB has an interesting way of organizing its modules and object files.
+This might be misleading, since the extensions are `.mod` and `.module`.
+In short, `.mod` files are the final binary that will be loaded at boot time, while `.module`s are intermediate unstripped object files that contains symbol and debug information.
+For the purpose of debugging, I needed the `.module` files.
+
+Obviously, the first module I looked into was `relocator.module`.
+I then needed to determine the base address where I had added my breakpoint.
+The plan was to:
+
+- print the address of the function using `__builtin_return_address(0)`
+- object dump the module to find the offset of the function
+- get the base address by substracting the offset from the address
+
+Pretty straightforward, right?
+Normally, yes.
+But here?
+Well, not quite.
+After performing this process with the printed address, `gdb` was showing code from `malloc_in_range` instead of `grub_relocator_alloc_chunk_align`.
+Additionally, it did not match the assembly (which was an infinite loop with a dummy value).
```

Comment:
>  (which was an infinite loop with a dummy value)

It would be nice to also explain what this means, e.g. explain how you would usually add manual breakpoints and such.